x-pack/libbeat/management/api/event_reporter.go

```diff
@@ -0,0 +1,131 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package api
+
+import (
+	"sync"
+	"time"
+
+	"github.com/joeshaw/multierror"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+var debugK = "event_reporter"
+
+// EventReporter is an object that will periodically send asyncronously events to the
+// CM events endpoints.
+type EventReporter struct {
+	logger       *logp.Logger
+	client       AuthClienter
+	wg           sync.WaitGroup
+	events       chan Event
+	period       time.Duration
+	maxBatchSize int
+}
+
+// NewEventReporter returns a new event reporter
+func NewEventReporter(
+	logger *logp.Logger,
+	client AuthClienter,
+	period time.Duration,
+	maxBatchSize int,
+) *EventReporter {
+	log := logger.Named(debugK)
+	return &EventReporter{
+		logger:       log,
+		client:       client,
+		period:       period,
+		events:       make(chan Event),
+		maxBatchSize: maxBatchSize,
+	}
+}
+
+// Start starts the event reported and wait for new events.
+func (e *EventReporter) Start() {
+	e.wg.Add(1)
+	go e.worker()
+
+	e.logger.Info("Starting event reporter service")
+}
+
+// Stop stops the reporting events to the endpoint.
+func (e *EventReporter) Stop() {
+	e.logger.Info("Stopping event reporter service")
+	close(e.events)
+	e.wg.Wait()
+}
+
+func (e *EventReporter) worker() {
+	defer e.wg.Done()
+
+	ticker := time.NewTicker(e.period)
+	defer ticker.Stop()
+
+	var buffer []Event
+	for {
+		select {
+		case event, ok := <-e.events:
+			if !ok {
+				e.reportEvents(buffer)
+				return
+			}
+			buffer = append(buffer, event)
+		case <-ticker.C:
+			e.reportEvents(buffer)
+			buffer = nil
+		}
+	}
```

Comment:
The buffer is unbounded, which means we have no control over memory usage. If one event producer becomes too chatty we either have to drop or block.

The channel itself is unbuffered. The effect of the channel being unbuffered will be back-pressure/blocking whenever we publish.

Alternative implementation without blocking:

```
type EventReporter struct {
  done chan struct{}

  buffer []Event
  mu sync.Mutex
}

func (e *EventReporter) AddEvent(evt Event) {
  e.mu.Lock()
  e.buffer = append(e.buffer, evt)
  e.mu.Unlock()
}

func (e *EventReporter) worker() {
  var done bool

  for !done {
    select {
      case <-e.done:
        done = true
      case <-ticker.C:
    }

    // flush buffer:
    var buf []Event
    e.mu.Lock()
    buf, e.buffer = e.buffer, nil
    e.mu.Unlock()
    e.reportEvents(buf)
  }
}
```

Instead of clearing the buffer one can use double-buffering and set `e.buffer = e.buffers[e.last][:0]`, so to keep the capacity.

Reply:
@urso I was wondering if we should instead use a buffered channel set with the limit of the batch size, which should keep memory under some limits, at most we would have 2 batches in memory.

Reply:
Another thing we would do could be to actually send events either on the number of events or the ticker trigger.

Reply:
The problem with the above is chatty clients could be a problem, in that case we might consider throttling somehow the events.

Reply:
Is there a notion of event priority? Maybe we could filter/remove some events if we need to make space.
I'm also afraid of potentially blocking the beat if we integrate this with logging.

Reply:
@urso Having priority sound good, lets do it when we integrate with logging.